### PR TITLE
3.10 Update - Massive Rewrite - Completed

### PR DESCRIPTION
**Pull request by tetrisdev**
In Progress Rewrite of the Entire mod to make it more maintainable.

TLDR:
Pulling out Changes into their own files.
Updating Hardcoded MongoIDs to use provided Enums by SPT to make updating in the future easier.
Better Logging in case of errors.

Explanation:
Added a ...Changer.ts for every type of change we want to make. This takes a dependency container and then with .apply() we apply the config options chosen by the user.

In apply() we find the individual methods that are then doing the changes.

Changed the config to .json5 with a better layout
Added types.ts for a Config Interface, so we have better typing regarding the config.

Moved all the whitelists, crafts, and larger changes into ./assets, this makes for more readable code.
Could also do this for the Barter changes with Traders...

Softcore.ts is the entrypoint to the mod and first loads the config and the wrapped logger.
Then in PostDBLoad() applies all the changes using the changers.

TODO: 
Add better Error Catching, currently there is none and that could lead to the server crashing... oups.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-	"name": "Softcore",
-	"version": "2.1.2",
-	"main": "src/mod.js",
+	"name": "softcore",
+	"version": "3.0.0",
+	"main": "src/Softcore.js",
 	"license": "MIT",
 	"author": "ODT",
-	"sptVersion": "~3.9",
+	"sptVersion": "3.10.3",
```
Comment (sapid) [inline package.json]: Changing this to `"~3.10"` will prevent the need for an update each time SPT gets a patch; this already breaks on 3.10.4.

File: config/config.json5 [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        alternativeCategories: true,
+        pacifistFence: {
+            // To go algong with the theme of this mod, Fence also sells only pacifist items. His prices are dependant on scav karma, so at 6 karma he will sell items almost at the same price Therapist buys them from you.
+            enabled: true,
+            numberOfFenceOffers: 30,
+        },
+        // Rebalances costs of hideout cases to fair and resonable prices and barters. Also changes LEDX dogtag barter
+        reasonablyPricedCases: true,
+        // EXPERIMENTAL option, makes Skier use Euros for all trades, and quest rewards, mostly just for fun and diversity. Adjusts assorts and loyalty levels accordingly. Turned off by default, because needs adjusting existing profiles salesSum (you can do it manually, find TradersInfo in your %profileID%.json, and in 58330581ace78e27b8b10cee key divide salesSum by 115 without leaving the remainder), unless you start a profile with this enabled.
+        skierUsesEuros: true,
```
Comment (sapid) [inline config/config.json5]: The comment indicates `skierUsesEuros` is disabled by default, but it is enabled by default. Leftover from testing, maybe?

File: package.json (hunk 1 above)
Comment (tetrisdev) [reply package.json]: Change has been made, thanks for the spot 👍

File: config/config.json5 (hunk 2 above)
Comment (tetrisdev) [reply config/config.json5]: Yes this is a leftover from testing. Updated, thanks 👍

File: src/changers/FasterbitcoinFarmingChanger.ts
Comment (sapid) [inline src/changers/FasterbitcoinFarmingChanger.ts:1]: I got an error due to the filename when running the SPT server in docker. I believe this wasn't detected earlier because the NTFS filesystem (Windows default) has no case sensitivity by default. On a case-sensitive filesystem like `ext4`, the lowercase `b` in `bitcoin` results in a failure to find this file when trying to import it. Changing the filename to `FasterBitcoinFarmingChanger.ts` fixed this problem for me.

File: src/changers/TraderChangesChanger.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+                    // Adjust all requirements' counts in this barter
+                    for (const requirement of traderAssort.barter_scheme[barterID][0]) {
+                        countAdjustment(requirement);
+                    }
+                }
+            }
+        }
+    }
+
+    private doSkierUsesEuros() {
```
Comment (sapid) [inline src/changers/TraderChangesChanger.ts:245]: It seems like this needs a check for converting the price on EUR. That is, if you try to buy EUR from Skier with this enabled, he will charge you in EUR. This makes buying EUR from him... complicated. 😎

image

**Review by sapid — CHANGES_REQUESTED** [review]

File: src/changers/FasterbitcoinFarmingChanger.ts
Comment (tetrisdev) [reply src/changers/FasterbitcoinFarmingChanger.ts:1]: I've been trying to see where i've made a case mistake. Searching for the term even without Case-Sensitivity i'm not seeing the spelling you have. Might have to do a new pull and rebuild on your end.

File: src/changers/TraderChangesChanger.ts (hunk 3 above)
Comment (tetrisdev) [reply src/changers/TraderChangesChanger.ts:245]: Fixed with ff7f93c
